### ISSUE TITLE
Implement Apple Pay recurring payment for iOS

### ISSUE THREAD

**Issue by BPerlakiH** [kiwix/kiwix-apple#1032]

**Comment by BPerlakiH** [comment]
@kelson42 @rgaudin 

I investigated this option. It seems that in order to effectively manage recurring payments / subscriptions, we need to authenticate our users (meaning sign in / sign up).
Once we have a user ID, we can via stripe, create a stripe user for them, save (in stripe) their default payment method (using stripe customer + setup intent), and create their subscriptions.

The other thing that the user needs to be able to do is to manage the subscription (eg. have the ability to resign). For this the sign in is also required.

In order to do this, we need to integrate with (the most supported one would be) [Sign in with Apple](https://developer.apple.com/documentation/authenticationservices/implementing-user-authentication-with-sign-in-with-apple).

**Comment by kelson42** [comment]
> @kelson42 @rgaudin 
> 
> I investigated this option. It seems that in order to effectively manage recurring payments / subscriptions, we need to authenticate our users (meaning sign in / sign up).
> Once we have a user ID, we can via stripe, create a stripe user for them, save (in stripe) their default payment method (using stripe customer + setup intent), and create their subscriptions.
> 
> The other thing that the user needs to be able to do is to manage the subscription (eg. have the ability to resign). For this the sign in is also required.
> 
> In order to do this, we need to integrate with (the most supported one would be) [Sign in with Apple](https://developer.apple.com/documentation/authenticationservices/implementing-user-authentication-with-sign-in-with-apple).

Considering that each Apple user has an AppleID, we should probably go forward with that.

**Comment by rgaudin** [comment]
I estimate a few days of my time. Let's say 4.

@Popolechien you'll have to decide if it's worth it and when to implement it

**Comment by Popolechien** [comment]
Ok let us put this on hold until 6 months after launch of the single-support option so we can assess the likelihood (and RoI) of getting recurring donors.

**Comment by Popolechien** [comment]
Pushing this back up to the top. @rgaudin Do we know how many individual donors went through the Apple donation system?

**Comment by Popolechien** [comment]
Thanks. Interesting that they would all pick the same amount.

I also do not see the last transaction in our Stripe log, so I assume it did not go through?

generally speaking it looks like Apple transactions are the only ones that do not have a customer name in the log, and by that metric I can only find another (10USD as well) donation in March. 

I guess that answers the question of whether implementing recurring donations is worth it.

**Comment by kelson42** [comment]
To ky opinion, at this stage, we have other priorities. But we could tackle side sometime in the future. Potential of recurring donation should not be underestimate.

**Comment by kelson42** [comment]
@Popolechien @BPerlakiH @rgaudin I would like to reconsider this issue considering the recent increase in usage and donations.

**Comment by rgaudin** [comment]
I've spent some time to look into this but it's still not 100% clear to me.

Here's what I think could work:

- Add Single/Monthly switch in UI. ATM there's a static `bool` in the code that toggles the wording to a _montly-oriented_ one but it's fake. It ends up charging a single time, I ~~lost~~donated $5 to confirm 😅
- Monthly UI should collect User's email address
- On confirmation, instead of `/payment-intent`, we'd call a new endpoint in our server that [creates a Customer](https://docs.stripe.com/api/customers/create) (using email address) and [a Subscription](https://docs.stripe.com/api/subscriptions/create) (using Customer, amount, currency and periodicity –monthly) and return the [PaymentIntent](https://docs.stripe.com/api/invoices/object#invoice_object-payments-data-payment-payment_intent) from the [LatestInvoice](https://docs.stripe.com/api/subscriptions/object#subscription_object-latest_invoice) -- [so](https://stackoverflow.com/questions/71916702/stripe-subscription-with-apple-pay)
- On `invoice.paid` event, we'd trigger an email to the Customer with details on how to manage subscription.
- We'd build a very basic webUI with:
  -  home: email field (“log-in to manage your subscription”)
  - validate with the OTP you just received via email
  - display subscription details: Start date, schedule, amount, list of payments
  - Cancel button that triggers email and [appropriate action](https://docs.stripe.com/api/subscriptions/cancel) at Stripe

Haven't tested but we could POC that easily (the Intent part for communication with the App).

Note that:

- It would be a Stripe Subscription, not an Apple subscription so it won't appear in Apple account.
- It would not be integrated with [our shop](https://get.kiwix.org/shop) at all.
- If it works as described above, we can spare the database and just query Stripe API for everything. Only the OTP thing would be in-app but that can stay in memory.
- With a bit more effort on the webUI, we could get rid of donorbox that we only keep because we need a cancel mechanism for recurring donations.
- This system is not be apple-pay specific: we're using Stripe's SDK and we configured Apple Pay for Stripe.
- Correlated to previous point, I believe there's no additional Apple-related requirement because Stripe's doing the heavy lifting of [keeping the DPAN/MPAN and triggering the MIT](https://docs.stripe.com/apple-pay/apple-pay-recurring).

**Comment by kelson42** [comment]
@rgaudin OK, but I really still wonder that so much work has to be homebrewed.

**Comment by rgaudin** [comment]
Well I don't think we can have both no-code and native integration (that we settled on using).
Also it's a very small amount of code (about 100l) that consists essentially in calling Stripe API.
We could have chosen to use Stripe completely and either pop-out of app for Web donation or include the Stripe UI SDK.

**Comment by BPerlakiH** [comment]
@kelson42 @rgaudin I also looked into this from the "Sign up/sign in with Apple ID".

So far it looks good, I've tested it, it works on my mac.
We can request the user full name and email (Apple even gives the user the option to hide the email, which still gives us a relay email, ending with "@privaterelay.appleid.com").
We need to store the full name and the email on the first successful attempt though as the name won't be returned later on.
We also get:
- a userID
- a valid JWT token signed by apple, which in itself includes, the email address, the userID, expiry (about 1 day).
it looks like this (I've deleted the middle parts with ...):
```json
{
  "iss": "https://appleid.apple.com",
  "aud": "self.Kiwix",
  "exp": 1778532260,
  "iat": 1778445860,
  "sub": "001737.e89.....1afe.0858",
  "c_hash": "IsX.....L5rWy3jA",
  "email": "w..k@privaterelay.appleid.com",
  "email_verified": true,
  "is_private_email": true,
  "auth_time": 1778445860,
  "nonce_supported": true,
  "real_user_status": 0
}
```

So I suppose with that we can put this together.
We just need to agree on how to do it. My thinking is the following:
1) we can cache the JWT token locally (on the client side)
2) if we have a JWT token and it's not expired, the user should be able to:
- subscribe to a monthly donation
- manage existing donations
3) if the JWT token is expired or we don't have one, the user can sign in
4) if login success and we have the name + email, we should register such user on the back-end, if we only get the JWT token but no name that is an already registered user and should be logged in

Let me know if it makes sense to you as well.


Also I think we should only trigger this, once the user selects the monthly donation option, otherwise it's not really required.
The strange part is that [there's no delete/sign-out option directly](https://stackoverflow.com/questions/58836179/how-to-sign-out-of-apple-after-being-authenticated/58841671#58841671). The user can remove the apple ID sign in only in the System Preferences (outside of the app). I suppose if the JWT token expires, that means the user is logged out...

**Comment by rgaudin** [comment]
Thank you @BPerlakiH I also think this sign in should only be for monthly and we should be careful in how we phrase it as it could confuse users to think it will be an App Store subscription.

**Comment by BPerlakiH** [comment]
This was the former UI we had for the monthly subscriptions:

<img width="432" height="512" alt="Image" src="https://github.com/user-attachments/assets/49488e5f-b27f-44ff-96ea-197501387ce1" />

After some experiments, I think the user flow should be first login, then select amount.

**Comment by kelson42** [comment]
> After some experiments, I think the user flow should be first login, then select amount.

We should try to avoid this login thing IMHO, nobody wants to give money AND have the hassle of an authentication.

Why not:
* People choose the amount
* we get the appleid email automatically
* we send an email with the URL to access the subscription mgmt page

**Comment by BPerlakiH** [comment]
@kelson42 you might be right.

I figured out a way to enforce user email to be attached to the payment, so hopefully that should be enough, right @rgaudin ?

Please see the recordings sent over on PM.

The Apple ID sign in is also ready, but I guess we won't need it if an email does work.

**Comment by rgaudin** [comment]
> [@kelson42](https://github.com/kelson42) you might be right.
> 
> I figured out a way to enforce user email to be attached to the payment, so hopefully that should be enough, right [@rgaudin](https://github.com/rgaudin) ?

Absolutely, if you can submit that email (or a private-relay one) to the server ; that would be enough and we'd stay on the original scenario with the added benefit that this info is gathered automatically and not entered.

> Please see the recordings sent over on PM.

LGTM

**Comment by rgaudin** [comment]
@BPerlakiH I have implemented the recurring donation (monthly) on the server.

You can take a look at the API: https://api.donation.kiwix.org/v1/

I've also pushed a commit on #1579 that extracts the email and send it along the `setup-intent` call.

It's working fine but there's a new UX problem: when that setup-intent calls succeed, at the moment, we are directly showing our thank you popup.
That was ~OK for single payment because while the actual charge had not happen at that time it would in the second to follow with very few risk to fail.
Now, when we return from setup-intent, there are still a few steps happening between stripe and our server API with higher chances to fail (due to bug of ours or payment-related things). It's all happening within 5s but our thank you popup is now too optimistic.
I think we should display some sort of loader and poll the API for a confirmation.

I haven't implemented that confirmation mechanism yet as I'd want us to discuss it first.
I am thinking it could return an ID along the secret (the `SetupIntent` one) and client would poll with it. Backend would retrieve the intent and check if it succeeded. If so, it'd find the `Customer` and look for a _related_ `Subscription` (then `Invoice`, then `Payment`) to return either `failed`, `succeeded`, `pending`.

Another very important detail: to prevent multiple subscriptions from user, if one already has an active subscription, the setup-intent call fails with a `409` response code.

Please also take a look at https://github.com/kiwix/donation-portal/issues/2

**Comment by BPerlakiH** [comment]
Thank you @rgaudin.
I did tested it out. One thing we should fix is the 409 error.
Currently it's not exactly as in the spec. It returns a 400 status error with the following json body:

<img width="1080" height="352" alt="Image" src="https://github.com/user-attachments/assets/7a28d98a-0925-4976-94ec-cd8dd2624456" />

I suppose it might be some typo on the back-end. 
I would prefer to rely on the status code itself (as per API docs), than trying to parse it out from the JSON.

**Comment by rgaudin** [comment]
Indeed! Thank you for catching it.
It's now fixed

```
│ INFO:     Monthly setup request for 500 USD reg@rskg.org                                                                                                   │
│ ERROR:    HTTPException(status_code=<HTTPStatus.CONFLICT: 409>, detail='Customer already has an active subscription')                                      │
│ INFO:     196.200.90.182:0 - "POST /v1/stripe/setup-intent HTTP/1.1" 409 Conflict
```

**Comment by BPerlakiH** [comment]
@kelson42 @rgaudin  Possible error message for this case:

<img width="432" height="310" alt="Image" src="https://github.com/user-attachments/assets/f60441e6-df1f-4e23-9dfd-26658b1b4d96" />

**Comment by BPerlakiH** [comment]
@rgaudin PR has been updated.

**Comment by BPerlakiH** [comment]
@kelson42 @rgaudin As I was adding new text to translations, it occurred to me that maybe it would be beneficial to localize the email instructions we are sending out as well...
Maybe can be moved under: https://github.com/kiwix/donation-portal/issues/2
Can be post MVP though...
I was thinking maybe we could send the language code used in the application to the server.

Speaking of which, we could potentially send the device type as well, which would cover https://github.com/kiwix/kiwix-apple/issues/1580

**Comment by rgaudin** [comment]
Absolutely. We can store the lang code on Stripe as well and have the webUI adapt also.

**Comment by BPerlakiH** [comment]
@rgaudin I've added these to the current "setup-intent" and "payment-intent" requests, so it can be parsed on the back-end:
```
{
  "currency": "USD",
  "locale": "en-US",
  "amount": 500,
  "email": "b****@***.com",
  "device": "mac"
}
```

The **locale** might come in the form of: `[language designator]-[region designator]`
I suppose we only need the first part before the "-".
As documented here:
https://developer.apple.com/documentation/foundation/nslocale/preferredlanguages
and under the LanguageIDs section here:
https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPInternational/LanguageandLocaleIDs/LanguageandLocaleIDs.html#//apple_ref/doc/uid/10000171i-CH15

The device will have one of the following values:
```
mac
iPhone
iPad
```

FYI: The emails and links are working as expected, I've tried 2 different accounts / providers.
Gmail for some reason puts the confirmation mail into the "spam" folder though...

**Comment by rgaudin** [comment]
OK, because we want to ultimately use this _service_ as a gateway for other/all donations, and for the sake of harmonization, I'd prefer we change to the following:

- rename `device` with `origin`.
- More specific origin maybe? Like `reader-iphone`? @kelson42 WDYT?
- rename `locale` to `lang`
- use only ISO-639-1 codes (`en`, `fr`)

> Gmail for some reason puts the confirmation mail into the "spam" folder though...

That might have to do with the wording of the email as technically we're OK I believe (we don't send emails, Mailgun guns) and we have properly configured SPF/DKIM/DMARC

**Comment by rgaudin** [comment]
I've implemented and deployed the changes to both staging and prod. See doc online.

**Comment by BPerlakiH** [comment]
> I'd prefer we change to the following:
> 
>     * rename `device` with `origin`.
> 
>     * rename `locale` to `lang`
> 
>     * use only ISO-639-1 codes (`en`, `fr`)

@rgaudin  I've updated the PR with those.

**Comment by kelson42** [comment]
Can I have a confirmation this issue is not only for iOS?

**Comment by BPerlakiH** [comment]
> Can I have a confirmation this issue is not only for iOS?

Yes, I can confirm it works for both iOS and macOS.
